ImperatorToCK3/Source/CK3/CK3World.cpp

```diff
@@ -245,22 +250,110 @@ void CK3::World::linkCountiesToTitleHolders(const ImperatorWorld::World& sourceW
+
 			if (capitalBaronyProvince >0) // 0 is not a valid province in CK3
 			{
-				auto owner = provinces.find(capitalBaronyProvince)->second->getOwner();
-				if (owner != "0")
+				if (provinces.find(capitalBaronyProvince)==provinces.end())
+					LOG(LogLevel::Warning) << "Capital barony province not found " << capitalBaronyProvince;
+				else
 				{
-					auto impMonarch = -1;
-					if (sourceWorld.getCountries().find(stoi(owner)) != sourceWorld.getCountries().end()) impMonarch = sourceWorld.getCountries().find(stoi(owner))->second->getMonarch();
-					if (impMonarch>=0) countyTitle->holder = std::to_string(impMonarch);
+					auto impProvince = provinces.find(capitalBaronyProvince)->second->srcProvince;
+					if (impProvince)
+					{
+						auto impMonarch = -1;
+						if (impWorld.getCountries().find(impProvince->getOwner()) != impWorld.getCountries().end()) impMonarch = impWorld.getCountries().find(impProvince->getOwner())->second->getMonarch();
+						if (impMonarch >= 0) countyTitle->holder = "imperator" + std::to_string(impMonarch);
+					}
+					else // county is probably outside of Imperator map
+					{
+						auto vanillaHistory = titlesHistory.popTitleHistory(name);
+						if (vanillaHistory) countyTitle->historyString = *vanillaHistory;
+					}
 				}
-				else // county is probably outside of Imperator map
+			}
+			titles.insert(std::pair(name, countyTitle));
+		}
+	}
+}
+
+
+void CK3::World::insertVanillaNonCountiesToTitles(const ImperatorWorld::World& impWorld)
+{
+	for (const auto& [name, landedTitle] : landedTitles.getFoundTitles())
+	{
+		if (name.find("c_") != 0 && name.find("b_") != 0 ) // title is a duchy or higher
+		{
+			bool toInsert = true;
+			// important check: if any of the title's de jure counties's holder is "0", don't insert the title
+			for (const auto& [vassalTitleName, deJureVassal] : landedTitle.getFoundTitles())
+			{
+				if (vassalTitleName.find("c_")==0 && titles.count(vassalTitleName) && titles[vassalTitleName]->holder == "0")
 				{
-					auto vanillaHistory = titlesHistory.popTitleHistory(name);
-					if (vanillaHistory) countyTitle->historyString = *vanillaHistory;
+					toInsert = false;
+					break;
 				}
 			}
-			titles.insert(std::pair(name, countyTitle));
+
+			if(toInsert)
+			{
+				auto vanillaTitle = std::make_shared<Title>();
+				vanillaTitle->titleName = name;
+				auto vanillaHistory = titlesHistory.popTitleHistory(name);
+				if (vanillaHistory) vanillaTitle->historyString = *vanillaHistory;
+				titles.insert(std::pair(name, vanillaTitle));
+			}
 		}
 	}
 }
+
+void CK3::World::linkSpouses(const ImperatorWorld::World& impWorld)
+{
+	auto counterSpouse = 0;
+	for (const auto& [ck3CharacterID, ck3Character] : characters)
+	{
+		std::map<int, std::shared_ptr<Character>> newSpouses;
+		// make links between Imperator characters
+		for (auto [impSpouseID, impSpouseCharacter] : ck3Character->imperatorCharacter->getSpouses())
```

Comment:
Never. Double looping is death.

Reply:
I don't see a problem here. Spouses map contains a few character ptrs in normal circumstances.

Reply:
I'm tired and seeing white mice. This isn't a double loop, pardon.